File: projects/06-offshore-wind-farm/backend/processor/transform.js
"use strict";

// The sort_key scheme (window_end # site_id) prevents turbine-1 and turbine-2
// readings for the same sensor_type in the same flush cycle from colliding
// on the DynamoDB primary key.
function buildSortKey(windowEnd, siteId) {
  return `${windowEnd}#${siteId}`;
}

function toItem(messageBody) {
  const payload = typeof messageBody === "string" ? JSON.parse(messageBody) : messageBody;
  const siteId = payload.site_id || "turbine-1";
  return {
    sensor_type: payload.sensor_type,
    sort_key: buildSortKey(payload.window_end, siteId),
    window_end: payload.window_end,
    window_start: payload.window_start,
    site_id: siteId,
    unit: payload.unit || "",
    count: payload.count,
    min: payload.min,
    max: payload.max,
    avg: payload.avg,
    latest: payload.latest,
    alerts: payload.alerts || [],
  };
}

module.exports = { toItem, buildSortKey };

File: projects/12-smart-building-energy/backend/processor/transform.py
import json


def to_reading_record(message_body):
    """Pure transform from a raw SQS message body (the fog node's window aggregate) into the flat DynamoDB record; sort_key is window_end + "#" + site_id so two floors flushing the same window don't collide on the range key."""
    data = json.loads(message_body) if isinstance(message_body, str) else message_body
    site_id = data.get("site_id", "floor-1")
    return {
        "sensor_type": data["sensor_type"],
        "sort_key": f"{data['window_end']}#{site_id}",
        "window_start": data["window_start"],
        "window_end": data["window_end"],
        "site_id": site_id,
        "unit": data.get("unit", ""),
        "count": data["count"],
        "min": data["min"],
        "max": data["max"],
        "avg": data["avg"],
        "latest": data["latest"],
        "alerts": data.get("alerts", []),
    }
